### PR TITLE
added map-node-sets->list macro

### PR DESCRIPTION
**Pull request by slyrus**
Added a little macro to map multiple node sets at once.

### CODE REVIEW

File: api.lisp [hunk 1]
```diff
@@ -68,6 +68,26 @@
...
+(defmacro map-node-sets->list (func &rest node-sets)
+  "@arg[func]{a function}
+   @arg[node-sets]{@class{node-set}}
+   @return{a list}
+   @short{Calls @code{func} for each tuple of nodes on the list @code{node-sets} and conses up
+   a list of its return values}
+
+   The operation is performed lazily, i.e. if it's terminated via
+   a non-local exit it doesn't necessarily cause the XPath engine to find
+   out all nodes in echo @class{node-set} in @code{node-sets} internally."
```
Comment (scymtym) [inline api.lisp]: Something went wrong here.

File: api.lisp [hunk 2]
```diff
@@ -68,6 +68,26 @@
         while (not (pipe-empty-p pipe))
         collect (funcall func (pipe-head pipe))))
 
+(defmacro map-node-sets->list (func &rest node-sets)
+  "@arg[func]{a function}
+   @arg[node-sets]{@class{node-set}}
+   @return{a list}
+   @short{Calls @code{func} for each tuple of nodes on the list @code{node-sets} and conses up
```
Comment (scymtym) [inline api.lisp]: This is a bit ambiguous. Maybe explicitly state that `func` is called with N arguments when `node-sets` consists of N node sets?

File: api.lisp [hunk 3]
```diff
@@ -68,6 +68,26 @@
...
+   a list of its return values}
+
+   The operation is performed lazily, i.e. if it's terminated via
+   a non-local exit it doesn't necessarily cause the XPath engine to find
+   out all nodes in echo @class{node-set} in @code{node-sets} internally."
+  (let ((pipe-syms (loop for node-set in node-sets
+                      collect (gensym "NODE"))))
+    `(loop ,@(loop for pipe-sym in pipe-syms
+                for node-set in node-sets
+                append `(for ,pipe-sym = (xpath::pipe-of ,node-set) then (xpath::pipe-tail ,pipe-sym)))
```
Comment (scymtym) [inline api.lisp]: Why the `xpath::` prefix?

File: api.lisp [hunk 4]
```diff
@@ -68,6 +68,26 @@
...
+        while (not (or ,@(loop for pipe-sym in pipe-syms
+                            collect `(xpath::pipe-empty-p ,pipe-sym))))
+        collect (funcall ,func ,@(loop for pipe-sym in pipe-syms
+                                 collect `(xpath::pipe-head ,pipe-sym))))))
```
Comment (scymtym) [inline api.lisp]: Indentation seems off in this line.

**Review by scymtym — CHANGES_REQUESTED** [review]
Minor changes requested in commit annotations.

File: api.lisp (hunk 1 above)
Comment (slyrus) [reply api.lisp]: Hmm... indeed. fixed!

File: api.lisp (hunk 2 above)
Comment (slyrus) [reply api.lisp]: Now reads: "Calls @code{func} with N arguments for each N-tuple of nodes
   on the N lists of @code{node-sets} and conses up a list of its
   return values"

File: api.lisp (hunk 3 above)
Comment (slyrus) [reply api.lisp]: Vestigal. Fixed.

File: api.lisp (hunk 4 above)
Comment (slyrus) [reply api.lisp]: Fixed. Thanks for the review!

**Review by scymtym — APPROVED** [review]

**Pull request merged**